project/JavalibIRCleaner.scala

```diff
@@ -0,0 +1,404 @@
+          else
+            ident
+        }
+      }
+    }
+
+    override def transformMemberDef(memberDef: MemberDef): MemberDef = {
+      super.transformMemberDef(memberDef) match {
+        case m: MethodDef =>
+          m.name match {
+            case name: Ident =>
+              MethodDef(m.flags, transformMethodIdent(name), m.args,
+                  m.resultType, m.body)(m.optimizerHints, m.hash)(m.pos)
+            case _ =>
+              m
+          }
+        case m =>
+          m
+      }
+    }
+
+    override def transform(tree: Tree, isStat: Boolean): Tree = {
+      implicit val pos = tree.pos
+
+      val preprocessedTree = tree match {
+        /* Rewrite `throw sjs.runtime.unwrapJavaScriptExeption(arg)` to
+         * `throw arg`, which is OK because the Java lang lib never throws
+         * wrapped JavaScript exceptions.
+         */
+        case Throw(Apply(_, LoadModule(ClassRef("sjsr_package$")),
+            Ident("unwrapJavaScriptException__jl_Throwable__O", _), arg :: Nil)) =>
+          Throw(arg)
+
+        case _ =>
+          tree
+      }
+
+      val result = super.transform(preprocessedTree, isStat) match {
+        case New(cls, ctor, args) =>
+          New(cls, transformMethodIdent(ctor), args)
+
+        case t: Apply =>
+          Apply(t.flags, t.receiver, transformMethodIdent(t.method),
+              t.args)(t.tpe)
+        case t:ApplyStatically =>
+          ApplyStatically(t.flags, t.receiver, transformClassRef(t.cls),
+              transformMethodIdent(t.method), t.args)(t.tpe)
+        case t: ApplyStatic =>
+          ApplyStatic(t.flags, transformClassRef(t.cls),
+              transformMethodIdent(t.method), t.args)(t.tpe)
+
+        case NewArray(typeRef, lengths) =>
+          NewArray(transformArrayTypeRef(typeRef), lengths)
+        case ArrayValue(typeRef, elems) =>
+          ArrayValue(transformArrayTypeRef(typeRef), elems)
+
+        case IsInstanceOf(expr, typeRef) =>
+          IsInstanceOf(expr, transformTypeRef(typeRef))
+        case AsInstanceOf(expr, typeRef) =>
+          AsInstanceOf(expr, transformTypeRef(typeRef))
+
+        case LoadJSConstructor(cls) =>
+          genLoadFromLoadSpecOf(cls)
+        case LoadJSModule(cls) =>
+          genLoadFromLoadSpecOf(cls)
+
+        case t: ClassOf =>
+          if (transformTypeRef(t.typeRef) != t.typeRef)
+            System.err.println(s"illegal Class(${t.typeRef}) at $pos")
+          t
+
+        case t =>
+          t
+      }
+
+      validateType(result.tpe)
+      result
+    }
+
+    private def genLoadFromLoadSpecOf(cls: ClassRef)(
+        implicit pos: Position): Tree = {
+      jsTypes.get(cls.className) match {
+        case Some(classDef) =>
+          classDef.jsNativeLoadSpec match {
+            case Some(loadSpec) =>
+              genLoadFromLoadSpec(loadSpec)
+            case None =>
+              System.err.println(s"$cls does not have a load spec at $pos")
```

Comment:
So this means it is not a native class, right? Maybe be more explicit about this?

Reply:
No it *is* a JS type (otherwise we wouldn't have gotten in the `case Some(classDef)`) but somehow it does not have a load spec. This shouldn't actually happen ever. It can only happen if the *compiler* did something wrong. I have expanded the error message to say so.